torchvision/prototype/features/_dataset_wrapper.py

```diff
@@ -0,0 +1,388 @@
+
+    node_type = _get_node_type(pytree)
+    flatten_fn = SUPPORTED_NODES[node_type].flatten_fn
+    child_pytrees, context = flatten_fn(pytree)
+
+    flat = []
+    for children_pytree, children_spec in zip(child_pytrees, spec.children_specs):
+        flat.extend(tree_flatten_to_spec(children_pytree, children_spec))
+
+    return flat
+
+
+class VisionDatasetFeatureWrapper(Dataset):
+    _wrappers_fns: Dict[Type[datasets.VisionDataset], PyTree] = {}
+
+    def __init__(self, dataset: datasets.VisionDataset, wrappers: PyTree) -> None:
+        self.vision_dataset = dataset
+        self.sample_wrappers = wrappers
+
+        # We need to disable the transforms on the dataset here to be able to inject the wrapping before we apply the
+        # transforms
+        self.transform, dataset.transform = dataset.transform, None
+        self.target_transform, dataset.target_transform = dataset.target_transform, None
+        self.transforms, dataset.transforms = dataset.transforms, None
+
+    def __getattr__(self, item: str) -> Any:
+        with contextlib.suppress(AttributeError):
+            return object.__getattribute__(self, item)
+
+        return getattr(self.vision_dataset, item)
+
+    def __getitem__(self, idx: int) -> Any:
+        # This gets us the raw sample since we disabled the transforms for the underlying dataset in the constructor
+        # of this class
+        sample = self.vision_dataset[idx]
+
+        wrappers_flat, spec = tree_flatten(self.sample_wrappers)
+        # We cannot use `tree_flatten` directly, because the spec of `self.sample_wrappers` and `sample` might differ.
+        # For example, for `COCODetection` the target is a list of dicts. To be able to wrap this into a dict of lists,
+        # we need to have access to the whole list, but `tree_flatten` would also flatten it.
+        sample_flat = tree_flatten_to_spec(sample, spec)
+        wrapped_sample_flat = [wrapper(item) for item, wrapper in zip(sample_flat, wrappers_flat)]
+        sample = tree_unflatten(wrapped_sample_flat, spec)
+
+        # We don't need to care about `transform` and `target_transform` here since `VisionDataset` joins them into a
+        # `transforms` internally:
+        # https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/datasets/vision.py#L52-L54
+        if self.transforms is not None:
+            sample = self.transforms(*sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        return len(self.vision_dataset)
+
+    @classmethod
+    def _register_wrappers_fn(cls, dataset_type: Type[datasets.VisionDataset]) -> Callable[[PyTree], PyTree]:
+        def register(wrappers_fn: PyTree) -> PyTree:
+            cls._wrappers_fns[dataset_type] = wrappers_fn
+            return wrappers_fn
+
+        return register
+
+    @classmethod
+    def from_torchvision_dataset(
+        cls,
+        dataset: datasets.VisionDataset,
+        *,
+        keep_pil_image: bool = False,
```

Comment:
This flag gives the user the option to keep images as PIL if they like. I've use `False` as default to promote the new functionality. In our experiments we saw that doing everything as tensors is on par with PIL regarding speed and accuracy.

Reply:
On a second thought, we don't want disrupt peoples workflows with this wrapper. So we should at least have `True` as default or remove the flag completely. Users can always put a PIL -> Tensor transform into their pipeline without any extra cost compared to doing it here.